Move Entities Saved States from Modal to Sidebar

Addison-Stavlo opened the pull request:
## Description
<!-- Please describe what you have changed or added -->
This PR explores moving the `entitiy-saved-states` component from a Modal to a Sidebar as discussed in issue #20421 .

This also enables applicable entities to be selected in the editor through the sidebar.

![sidebar-preview](https://user-images.githubusercontent.com/28742426/78937766-05810c00-7a7f-11ea-9910-cb155bc1c375.gif)

<hr/>

Notes on design parity (or lack thereof) and future goals:
* Checkbox Controls -  the Sidebar designs in the issue have gotten rid of the checkbox controls.  We have kept them in this iteration since it would not make sense to remove them before its replacement functionality (such as a "discard changes" for each entity) has been implemented.  I am having problems figuring out how to do this with the current tools available in core data, and this will be the next step to look into following this PR.  
* "Select Entity" Button - Similarly, once we are able to get rid of the Checkbox controls,  the 'Select Entity' button would be removed and its functionality would be added to the entity Title as shown in the design figmas.
* Panel Body / dropdowns - Temporary solution to keep things decently styled and well organized.  These would be removed and refactored to match the designs once the functionality for other options shown is available ( 'discard changes', 'save as', etc.)
* Icons - A handful of Icons shown in the Figma designs are not yet available in our Icons package.  I have temporarily made a 'best guess' using available icons as placeholders until these can be updated.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
Tested on local docker setup in the various post editors and site editor.

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Addison-Stavlo committed:
moved to sidebar, functional in post editor / hidden in site

Addison-Stavlo committed:
copied over some styles, working in site editor now

Addison-Stavlo committed:
added close button

Addison-Stavlo committed:
added icons and styles

Addison-Stavlo committed:
added selection feature

Addison-Stavlo committed:
refactored store goo

Addison-Stavlo committed:
refactored z-index

Addison-Stavlo committed:
refactored entities saved states

Addison-Stavlo committed:
refactored scss

Addison-Stavlo committed:
added test for actions

Addison-Stavlo committed:
added tests for reducer

Addison-Stavlo committed:
added selector test

Addison-Stavlo committed:
fixing lint

Addison-Stavlo committed:
fixing more lint

Addison-Stavlo committed:
cleanup bad auto-pretty

Addison-Stavlo committed:
last of cleanup?

Addison-Stavlo committed:
renamed select entity button

youknowriad reviewed (commented):
Feature wise, this makes sense and I'd love to land an iteration of this but from the code perspective, I have some concerns, the main ones are:

We shouldn't be relying on the "editor" package ini "edit-site". The editor package is a post-specific one. We might need some refactoring to achieve this though. 

The "edit-post" package also has the concept of "pre-publish" panel, I'm not sure we can share code (at least as is) with the "edit-site" package since the logic for that panel is buried somewhere in the editor package, but they should at lease behave closely. I wonder if there are some generic components to extract while leaving the specifics to each top level package (edit-*).

Addison-Stavlo commented:
>  We shouldn't be relying on the "editor" package ini "edit-site". The editor package is a post-specific one. We might need some refactoring to achieve this though.

Ah I see!  `entities-saved-states` and its associates store are all setup in `editor` and is then used inside both `edit-post` and `edit-site`, which is why I was assuming it wasn't `edit-post` specific.  Do you think refactoring all of this over to `block-editor` package would make the most sense?

> the logic for that panel is buried somewhere in the editor package, but they should at lease behave closely. 

Thats essentially what this is modeled after (mostly from the scss) with a +1 on the z-index so it plays well with the publish panel inside the post editor as well.

youknowriad commented:
Just noticed that "EntitiesSavedState" is on Editor which doesn't make a lot of sense to me. I think it should probably be in "core-data" since it's related to "entities". That said I'm fine with keeping it there in this PR and solving that part separately.

What I'd recommend is to try to build this PR without introducing any editor specific things (core/editor store or components) into EntitiesSavedState, potentially, by using props for the component.

It's going to be a challenge to try to reuse things between edit-site and edit-post. block-editor is not going to be the right place for these when they deal with WordPress specifics and APIs.

Addison-Stavlo commented:
> What I'd recommend is to try to build this PR without introducing any editor specific things (core/editor store or components) into EntitiesSavedState,

It looks like the only `core/editor` store the entities-saved-states is currently using are the actions/selectors I just added for controlling the open/closed state.  If we were to move this to core-data, these would go along with it and Im assuming that would be fine?  Or is there some other issue here?

Then our edit-site would call core-data instead of the editor package and we wouldn't be conflicted there.  And the edit-post package would continue to call core/editor for editPost/savePost where it does and would move to using core-data for the saved states sidebar.

Does the above sound good?  Or is there some other conflicting piece I am missing with this?

youknowriad commented:
>  If we were to move this to core-data, these would go along with it and Im assuming that would be fine? Or is there some other issue here?

We shouldn't move UI related state to core-data. UI related things should stay in the UI related packages (edit-post, edit-site). I think we should use props. `onSelectEntity` or things like that..

Addison-Stavlo commented:
> We shouldn't move UI related state to core-data. UI related things should stay in the UI related packages (edit-post, edit-site). I think we should use props. onSelectEntity or things like that..

Gotcha.  For the Modal, we were using props and passing down to `entities-saved-states` as a child component of the save/publish buttons.  This works well for a modal being fixed position across the entire screen, but for a sidebar we need to place the component elsewhere for proper dom hierarchy and styles.

Modeling functionality around how the publish panel works, instead of `entities-saved-states` being a child of the top-bar buttons, it has been moved to the `actions` prop of the layout.  ( layout->index.js in edit-post, and editor->index.js of edit-site).  For us to pass the correct props in this location it seems like the simplest thing to do would be to use the store to bridge that gap?  In which case, do we want 1 copy of these actions/selectors in 1 edit-post store and another copy of essentially the same thing in the edit-site store?  Is there not any package where we could use these same actions/selectors for both uses?

youknowriad commented:
> do we want 1 copy of these actions/selectors in 1 edit-post store and another copy of essentially the same thing in the edit-site store?

I think that's totally fine, we shouldn't be afraid of duplicating code. Sometimes I feel early abstractions do more harm than good. Now later, if we want to reuse more stuff, maybe the "interface" package is a good fit since it's a "WPAdmin UI" package but I'm still not sure about the APIs at this point so better to start simple.

Addison-Stavlo committed:
added actions/reducers/tests to edit-post

Addison-Stavlo committed:
added store things to edit-site

Addison-Stavlo committed:
restructured to run on isOpen/closePanel props

Addison-Stavlo committed:
removed from editor package store

Addison-Stavlo commented:
@youknowriad  - That makes sense to me.  I removed the store stuff from the editor package and moved them to edit-post and edit-site instead.  The `entities-saved-states` component now receives these as props instead of relying on one specific store.  

After taking a quick look at moving `entities-saved-states` over to `core/data` package:  it looks like the package doesn't currently have any components exported.  Would it make sense to add a components directory and the required dependencies to export components from this package, or maybe it will make more sense to find a home for it elsewhere?  (we can move it later as you suggested, but I wanted to take a quick look).

> maybe the "interface" package is a good fit since it's a "WPAdmin UI" package but I'm still not sure about the APIs at this point so better to start simple.

Maybe "interface" would make sense as the home for both the saved states component and its associated store in the end? 🤔   Once we are more sure about the APIs and want to consolidate/refactor this around more.

youknowriad commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -69,6 +70,7 @@ function Layout() {
 		previousShortcut,
 		nextShortcut,
 		hasBlockSelected,
+		isEntitiesSavedStatesOpen,
```
We already have a "pre publish panel" on edit-post/editor, I feel this should just use the same selectors/controls. Or even render `EntitiesSavedStates` on the existing PrePublishPanel instead of having a separate panel.

Addison-Stavlo replied:
Im not sure I agree.

The pre-publish panel is only currently used in the pre-published state of the post editor.  The multi entity saving panel will need to be used regardless of whether or not the post is published.  If we wanted to use the same selectors/controls or the same panel, we would need to do some refactoring of the pre-publish panel itself to play nice and make sense in both pre and post publish modes.  As things sit right now, the separate panel both plays nicely with the existing pre-publish panel, allows for separation of concerns (entity saving vs. publishing), and keeps a parity of interface between multi entity saving in the site and post editors.

youknowriad replied:
> The pre-publish panel is only currently used in the pre-published state of the post editor

For the user, it's just the same pre-publish panel filled with different UIs depending on the state of its current changes.

> If we wanted to use the same selectors/controls or the same panel, we would need to do some refactoring of the pre-publish panel itself to play nice and make sense in both pre and post publish modes

Would that be an issue, I mean why is that not a good approach? 

>  As things sit right now, the separate panel both plays nicely with the existing pre-publish panel, allows for separation of concerns (entity saving vs. publishing)

I don't think this separation as a necessary one, for me, the user just adds and edits things in its content, if it happens to be things saved to multiple entities (like adding a site title block inside its post content), the pre publish panel adapts to clarify what is being saved. That said, I'd be happy to have more design thoughts here @jasmussen @MichaelArestad

jasmussen replied:
I'm very late to this one, and honestly I missed the part where #20421 landed on a sidebar solution. Which I personally feel like a sidebar is a wrong solution, exactly because it mimics the pre-publish flow as to be mistaken for it, when it is in fact quite different. 

The pre-publish flow is designed around providing a last minute checkup to _publishing new pages and posts_: surfacing for you things like tags or the permalink in case you dismissed the sidebar when you wrote. It even surfaces once more as a sanity check things like Jetpack publicize to let you know one last time before you publish that "hey — I'm about to ping every social network you have connected". It's an "are you sure you want to publish this post or page"?

Saving multiple template parts, also needs a pre-publish check but for a different reason that has nothing to do with tags, or pinging social networks. You may have inadvertently made changes to the entire website, so you need to be sure that's what you actually want to do. And because saving multiple templates at once is inherently a much more complex task than publishing a post, and making the two look the same does not feel like the right solution. For example the pre-publish flow has been optimized to allow you to "double-click" the publish button to _go fast_. But when you're saving multiple template parts, you really don't want to go fast.

So to the questions at hand there are two opposing views:

> As things sit right now, the separate panel both plays nicely with the existing pre-publish panel, allows for separation of concerns (entity saving vs. publishing)

.. and:

> I don't think this separation as a necessary one, for me, the user just adds and edits things in its content, if it happens to be things saved to multiple entities (like adding a site title block inside its post content), the pre publish panel adapts to clarify what is being saved.

Based on the premise that I think the foundation is wrong, it's hard to pick either. In fact this very stacking of sidebars on top of each other feels like an example of the kind of confusion of different saving contexts, that I believe we should avoid. 

However given a consensus was reached #20421, I will simply note this disagreement and suggest that of the two approaches, the one that at least keeps the panels different, is the one that I would go with, even if it means stacking sidebars.

Also worth noting, going that route, it means you'll need to add an additional on-focus tab stop for screen reader users to the one we already have:

![tabbing](https://user-images.githubusercontent.com/1204802/79461952-27d2d680-7ff7-11ea-83a9-aae6e67d3718.gif)

Addison-Stavlo replied:
> Would that be an issue, I mean why is that not a good approach?

Refactoring it to play nice isnt an issue, but the reasoning is a bit confusing.  We have a pre-publish panel designed to be used in the pre-published state... updating it to also work in a state where its underlying purpose is no longer relevant seems odd.  Especially when the motivation to do so is to add functionality that is not necessarily related to publishing.

And kind of going with your comment from earlier:
> we shouldn't be afraid of duplicating code. Sometimes I feel early abstractions do more harm than good.

Publishing and entity saving do function separately, and it does feel like trying to couple them together in one editor, while having a separate panel for the other, could unnecessarily make things more complicated and things more difficult for others to work with down the road?

Im not against doing the work to make it happen, just a bit concerned and confused about the reasoning.  

I'd like to hear what design thinks about sharing entity saving and publishing options in the same panel in the pre-publish state.  It seems like it could really clutter the panel as we converge on implementing more designs and functionalities as discussed in #20421, especially when there are a handful of entities in use.

Addison-Stavlo replied:
>  it means you'll need to add an additional on-focus tab stop for screen reader users to the one we already have:

Im not sure I follow.  The 'one we already have' as in the pre-publish panel doesn't work with screen readers currently?  Or something about adding this extra panel breaks that functionality?  I'm not super experienced with screen readers so I don't completely understand what is needed.

From a tab-stop accessibility perspective, it seems like both panels still work well.  The only conflicting issue I see is when the entity saving panel is open, you can tab down to 'open the publish panel', which will open the publish panel hidden behind the saving panel.  Saving through the save panel triggers the publish panel to open right afterwards, so maybe it would be best to disable/hide that button when the saving panel is open?  The other thought is to have it close the save panel to allow viewing of the publish panel, but then you are in the publish panel with unsaved entities which doesn't really make sense.  🤔

youknowriad replied:
Maybe that's one of those areas where it's hard to agree on something.

Noting here that it's possible to do both things at the same time "publish a post" and "save multiple entities", for that reason I find it confusing to offer multiple UIs to the user, I'd prefer some kind of two steps wizard but in the same UI. 

Now technically, the current publish panel already has the required a11y work to make it work ( a hidden button that shows up when you tab through the UI and the corresponding area-region), using a different panel, means another aria-region, another hidden button to open that specific panel. I feel that's wrong, it's like we're continuing to add new panels to the UI, they are visible to screenreader users. Imagine if they were visible to all users having and you have to see all these things at the same time? I also fear the bugs this might produce, all the logic about which panel to show/hide can become complex.

Addison-Stavlo replied:
Those are some good points @youknowriad.

With the current flow:  when there are multiple entities requiring saving, the 'Publish' button opens the entities saved states (either modal or sidebar) instead of the pre-publish panel.  On save, the supplied callback function then triggers the publish panel to open.  At this point, we only have 1 panel open at a time.  

However, if we do press the hidden button you mention, it opens the publish panel regardless of whether or not there are other entities requiring saving.  From here if we try to 'publish' it will open the save panel on top of the publish panel (if it wasnt already open when pressing the hidden button).  The most simple solution I can think of at the moment would be to make that hidden button behave the same way as the 'publish' button at the top (shouldn't it be doing that anyways?) and open the save panel first if necessary.  This way we should never have both sidebars open at the same time, I think the logic would be fairly simple and essentially mirror what is already going on within the standard 'publish' button. 🤔

Addison-Stavlo replied:
I am starting to see some advantages with sharing the same panel.  However I do wonder:

> I also fear the bugs this might produce, all the logic about which panel to show/hide can become complex.

If we were sharing the same panel, wouldn't the logic about which content to show/hide within the panel be just as complex?

youknowriad replied:
> If we were sharing the same panel, wouldn't the logic about which content to show/hide within the panel be just as complex?

I think it will be simpler personally because of the fact that these two panels show at the same time exactly if you see it this way:

 1- Entities is the first step in the pre-publish wizard
 2- Current pre-publish is the second one

the panel just opens when you click the button on the toolbar (or the hidden button)
The panel closes when you finish the saving

That said, that's the theory, may I propose that if we do a refactor, we do it as a separate branch/PR to compare?

Addison-Stavlo replied:
> That said, that's the theory, may I propose that if we do a refactor, we do it as a separate branch/PR to compare?

Certainly, I was thinking the same.  Il branch off of this one to explore refactoring around sharing the panel.

Addison-Stavlo replied:
@youknowriad  - After looking into this a bit more, I think the below solution will make the most sense (commit 1f52b14).  This change to the actions section of the Layout component will ensure that only 1 panel is rendered at any given time, so it should solve our concern about overlapping panels and screen readers with very simple logic.  This allows all behavior to remain the same without rendering any overlapping components.

I thought of the above while looking into refactoring the post-publish-panel to share the saving functionality.  Doing the latter seems to be a fairly complex refactor with much more room for errors in logic for desired functionality.  I don't think it makes sense given how we are currently set up, especially if the above solution will solve our concerns.

<hr/>

I have some lingering questions/concerns regarding the hidden 'Open publish panel' button given questionable behavior **after the post is in the published state** (even on the current master branch w/o the experiment enabled), maybe @jasmussen would have some insight here:
* After a post is in the published state, this hidden 'Open publish panel' button is still present.  Should it be?  There does not seem to be another way to open the panel in this state, so I'd assume the 'hidden' accessibility button shouldn't be present either?
* If it is supposed to sill be present at this stage:  It only actually opens the panel if the post is not dirtied.  If selected after changes have been made, it does not open the panel when clicked.  I assume this is a bug, or is there something else I am missing?

I assume we want this style of 'hidden' button to open the save panel when applicable both in edit-post as well as edit-site?  I could just use some confirmation about how the current implementation is supposed to work though, so I can set things up with the desired functionality in the post editor.

Addison-Stavlo committed:
created Actions component for post editor layout

Addison-Stavlo committed:
added save panel hidden button to post editor

Addison-Stavlo committed:
added hidden button to site editor

Addison-Stavlo committed:
added styles and class names

Addison-Stavlo committed:
fixed z-index name issue

Addison-Stavlo committed:
restored z-indexes to previous values, no need with non-overlapping panels

Addison-Stavlo commented:
I added some updates on the post-editor side so that:
*  actions panels will never overlap.
*  a tab accessible "open save panel" button appears when applicable
*  'hidden' a11y buttons will not overlap or show when panels are open
*  default behavior on the 'open publish panel' button remains the same.

And on the site-editor side, I added the "open save panel" a11y button.

youknowriad commented:
Just noting that I'm not be able to review this on the next couple days. @epiqueras you might be interested in this PR too.

Addison-Stavlo committed:
refactor ActionsPanel, always mount entitites"

Addison-Stavlo committed:
removed unnecessary vars and imports

Addison-Stavlo committed:
make save component always mounted in edit site"

Addison-Stavlo committed:
added some comments

Addison-Stavlo commented:
Rebased, and refactored the `<ActionsPanel />` component a bit and moved it from layout (packages/edit-post/src/components/layout/actions-panel.js).

What do we think about the state of this PR?  
* Without the experiment enabled everything should behave as it does currently.  
* With it enabled we will have order of precedence of Save Panel -> Publish Panel -> Save a11y Button (if applicable) -> Publish a11y Button - without having any overlapping panels or buttons. 
* The site editor will now have the Save a11y button set up similarly to how the publish one currently works.

Addison-Stavlo committed:
fixed closePanel for selecting entity in small width

epiqueras commented on packages/edit-site/src/components/editor/index.js:
```diff
@@ -68,6 +80,8 @@ function Editor( { settings: _settings } ) {
 
 	const inlineStyles = useResizeCanvas( deviceType );
 
+	const openSavePanel = useCallback( () => openEntitiesSavedStates(), [] );
```
This `useCallback` is redundant. `openEntitiesSavedStates` doesn't change.

epiqueras commented on packages/edit-site/src/components/editor/style.scss:
```diff
@@ -0,0 +1,23 @@
+.edit-site-editor__toggle-save-panel {
+	z-index: z-index(".edit-site-editor__toggle-save-panel");
+	position: fixed !important; // Need to override the default relative positionning
```
```suggestion
	position: fixed !important; // Need to override the default relative positioning
```

Addison-Stavlo replied:
Ah yeah, thats what I get for copying styles over from the other a11y buttons and only vaguely reading the comments.

epiqueras commented on packages/edit-site/src/components/editor/style.scss:
```diff
@@ -0,0 +1,23 @@
+	width: $sidebar-width;
+	background-color: $white;
+	border: 1px dotted $light-gray-500;
+	height: auto !important; // Need to override the default sidebar positionnings
```
```suggestion
	height: auto !important; // Need to override the default sidebar positionings
```

epiqueras commented on packages/edit-site/src/components/save-button/index.js:
```diff
@@ -52,9 +51,11 @@ export default function SaveButton() {
+	const { openEntitiesSavedStates: openSavePanel } = useDispatch(
+		'core/edit-site'
+	);
+	const openPanel = useCallback( () => openSavePanel(), [] );
```
This `useCallback` is redundant.

epiqueras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -6,32 +6,100 @@ import { some, groupBy } from 'lodash';
+		);
+		// Get parents of the entity's first block.
+		const parents = select( 'core/block-editor' ).getBlockParents(
+			blocks[ 0 ]?.clientId
```
What happens if the entity has no blocks? E.g. an empty template part.

Addison-Stavlo replied:
I guess a couple things regarding that:

An empty template part doesn't seem to trigger as being dirty until it has a block.  It seems clicking on the new template part initiates the paragraph block by default, but at this point it is no longer empty.

If that empty template part WERE to be dirty (hypothetically), this would pass `undefined` into getBlockParents which would then return an empty array.  From there since we don't have the parentId, the 'Select Entity' button would not be present.  I suppose this behavior would make sense, as if the template part has no blocks selecting it would not really make sense.  🤔

epiqueras replied:
That makes sense. Ideally, we would select the wrapper. This could happen when someone completely clears a template part. It's a super edge case, so let's not worry about it for now.

epiqueras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -6,32 +6,100 @@ import { some, groupBy } from 'lodash';
+	}, [] );
+
+	const { selectBlock } = useDispatch( 'core/block-editor' );
+	const selectParentBlock = useCallback(
+		() => selectBlock( parentBlockId ),
+		[]
+	);
+
+	const selectAndDismiss = useCallback( () => {
+		selectBlock( parentBlockId );
+		closePanel();
+	}, [] );
```
`parentBlockId` can change and should be a dep here.

epiqueras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -6,32 +6,100 @@ import { some, groupBy } from 'lodash';
 	);
 
+	// Set icon based on type of entity.
+	const { name } = firstRecord;
+	let icon = blockDefault;
+	if ( name === 'site' ) {
+		icon = layout;
+	} else if ( name === 'page' ) {
+		icon = page;
+	} else if ( name === 'post' || name === 'wp_template' ) {
+		icon = grid;
+	} else if ( name === 'wp_template_part' ) {
+		icon = blockDefault;
+	}
+
```
A constant enum will scale better.

Addison-Stavlo replied:
Would that be as with a `switch` case?  My apologies if I'm not understanding.

epiqueras replied:
```js
const ENTITY_NAME_ICONS = {
	site: layout,
	page: page,
	post: grid,
	wp_template: grid,
};
const icon = ENTITY_NAME_ICONS[ name ] || blockDefault;
```

Addison-Stavlo replied:
Oh yes!  That definitely seems like the cleanest and most scalable way to define this.  Thank you!

epiqueras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -106,40 +178,49 @@ export default function EntitiesSavedStates( { isOpen, onRequestClose } ) {
-						/>
-					);
-				} ) }
+	const dismissPanel = useCallback( () => closePanel(), [] );
```
Redundant `useCallback`.

Addison-Stavlo replied:
I think this is still necessary in that in the post editor `closePanel` fires the callback added in the publish button.

If we aren't specific about firing it with no arguments, then the event object will be passed in where we would expect to see `savedEntities` and cause an error.  I suppose we could define this without using `useCallback` since the dependency array is empty, but doesn't `useCallback` improve performance via memoization here anyways?

epiqueras replied:
Ahh, that makes sense. Let's add a comment to make it clear.

epiqueras reviewed (changes requested):
Nice work!

I made some minor comments.

My only overarching concern with this PR is that we are unnecessarily adding public API methods, selectors and actions, for something that is much more suited for local state. Both `Layout` and the `edit-site` `Editor` can handle the opening and closing of the panel with local state.

I also think we should standardize the terminology a bit. There are both references to "entities saved states" and "save panel". This can be very confusing for someone who hasn't followed along.

Addison-Stavlo commented:
Hi @epiqueras , thank you for taking an in depth look here.  I will work at addressing many of the issues you have noted.

A ❓  in the meantime:

> for something that is much more suited for local state. Both Layout and the edit-site Editor can handle the opening and closing of the panel with local state.

We have the a11y buttons in the Layouts that trigger the panel, but we also have the save button for each editor that will consume this state as well to trigger the panel when necessary.  This is why I ended up using the store for a centralized state, but do you think it would make more sense to redefine the store in Layout's component state and prop pass what is necessary down through the header to the save buttons?   I could refactor it to do this if that seems to make the most sense.

Addison-Stavlo commented on packages/base-styles/_z-index.scss:
```diff
@@ -71,6 +71,12 @@ $z-layers: (
+	".entities-saved-states__panel": 100001,
+	// For larger views, the wp-admin navbar dropdown should be at top of
+	// the multi-entity saving sidebar.
+	".entities-saved-states__panel-break-medium": 99998,
```
Oh gotcha!  I will look to setting it up that way, I had been following the example from the post-publish panel:

https://github.com/WordPress/gutenberg/blob/98f987cdccd77685c9f2ce5bc4fb83f28a3349d3/packages/base-styles/_z-index.scss#L72

Addison-Stavlo replied:
I updated this (as well as the publish panel) to use this convention.

Addison-Stavlo commented on packages/edit-post/src/components/layout/actions-panel.js:
```diff
@@ -0,0 +1,102 @@
+		};
+	}, [] );
+
+	const openSavePanel = useCallback( () => openEntitiesSavedStates(), [] );
```
I was necessary due to the callback function you pointed out that shouldn't be in state.  Since I will be removing that, I can remove this as well.

epiqueras replied:
Even with that,

```js
useCallback( () => openEntitiesSavedStates(), [] ) === openEntitiesSavedStates
```

Addison-Stavlo replied:
I think in this case if we just pass `openEntitiesSavedStates` on click, then the event object gets passed into it and then set as that callback function.  So the importance here is really the fact that we are specifically telling it to fire with no argument as opposed to the memoization or anything else.

Addison-Stavlo commented on packages/edit-post/src/components/layout/actions-panel.js:
```diff
@@ -0,0 +1,102 @@
+
+	// It is ok for these components to be unmounted when not in visual use.
+	// So we can use them in a return cascade since only 1 is to be present at a time.
+	const Unmountables = () => {
```
Ah good point.  

> A simple variable would suffice.

Do you mean to have a variable to use in the return statement with ternaries?  Or something else?

Do you think it would be suitable (if still necessary) to define this component standalone as opposed to inline inside the parent?

epiqueras replied:
Something like:

```jsx
export function Component( { x, y, z } ) {
	let content;
	if ( x ) content = <span>A</span>;
	else if ( y ) content = <span>B</span>;
	else if ( z ) content = <span>C</span>;
	return <div>Letter: { content }</div>;
}
```

Addison-Stavlo replied:
Thank you!  That makes sense here 😄

Addison-Stavlo commented on packages/edit-post/src/store/reducer.js:
```diff
@@ -238,6 +238,37 @@ export function deviceType( state = 'Desktop', action ) {
+			return {
+				...state,
+				isOpen: true,
+				onRequestClose: action.onRequestClose || function() {},
```
I tried that initially and encountered some errors.  But if we are removing the callback from the reducer (or the store changes altogether) this will be gone.

epiqueras replied:
It might be because `this` is set to the action object?

epiqueras commented:
Yeah, this seems like a good candidate for just props unless we see a need for this to be in the store.

>but now that the main component is now in the actions panel we need a way to get that callback to the component.

It can be done the same way. `Layout` can hold it in state.

Addison-Stavlo committed:
removed entities actions/reducers from edit-site editor

Addison-Stavlo committed:
passed localstate down through props in edit-site

Addison-Stavlo committed:
removed entities goo from edit-site store

Addison-Stavlo committed:
added parentBlockId as dep to callback in entitiy selection

Addison-Stavlo committed:
converted entities to local state in edit-post layout

Addison-Stavlo committed:
set up edit-post save panel on local state props

Addison-Stavlo committed:
removed entities goo from edit-post store

Addison-Stavlo committed:
fixed z-index naming convention

Addison-Stavlo committed:
refactored actions panel

Addison-Stavlo committed:
fixed comment misspellings

Addison-Stavlo committed:
refactored entitiesSavedStates icon enum

Addison-Stavlo commented:
Thanks again for your help with this @epiqueras !  

I think I have taken care of most of the suggestions, with a couple conversations left open.

The last thing I am not sure about is whether or not it makes sense to have the a11y 'Open Save Panel' button set up to use the same callback that the post-publish-button sets.  Currently, the post-publish-button sets our callback when clicked.  

Would we want to refactor that to set the callback before then, so that going through the flow with the a11y button fires the same callback when closing?  Or does it not really make sense since the 'publish' button and the 'open save panel' button are not really 1-1 in use and expectation?  All saving seems to happen regardless of missing the callback, and that flow doesn't seem incomplete. 🤔

Addison-Stavlo committed:
removed unnecessary comments

Addison-Stavlo committed:
added comment to redundant looking useCallback

epiqueras commented:
The post publish button needs to set that callback because it needs to open the publish panel after saving happens conditionally.

The a11y button doesn't need to open any panels after saving, right?

Addison-Stavlo commented:
Yeah, if thats the only concern then I think we are good.

epiqueras commented:
Nice!

`packages/e2e-tests/specs/experiments/multi-entity-saving.test.js` needs to be updated and we should be good to go.

Addison-Stavlo commented:
> packages/e2e-tests/specs/experiments/multi-entity-saving.test.js needs to be updated and we should be good to go.

Is there anything specific we want to update here?  Should we be adding tests for the a11y button here?

Luckily our selectors for the tests remained the same and assertions still make sense, so everything is still passing.

Addison-Stavlo committed:
updated description titles in test to no longer say 'modal'

epiqueras commented:
Let's just make sure that the two different a11y buttons and the two different panels render under the right circumstances.

Addison-Stavlo committed:
added tests for panels/a11y buttons rendering

Addison-Stavlo committed:
minor name and descriptor changes

Addison-Stavlo commented:
Those tests should be in place.  Travis had an error on downloading wordpress, so we will see if its happy after this next run.

Addison-Stavlo commented:
Hrm, now there is a test failing from `post-visibility`.  I don't think I have made any changes related to that, it seemed to be passing last night after adding the tests 🤔

epiqueras commented:
It can't find the post publish button.

https://travis-ci.com/github/WordPress/gutenberg/jobs/325144924#L1015

Addison-Stavlo commented:
> It can't find the post publish button.

Right...  Whats odd is that test was passing after all functional changes had been made on 0797c0c, it seemed to be passing as well after we added the tests on 545c200, and as of this morning (and the only update made changing a variable name in a localized test funciton) it started failing.

When I try to run it locally it fails on React API warnings regarding componentWillRecieveProps and legacy API use detected.

epiqueras commented:
Set `SCRIPT_DEBUG` to false.

Addison-Stavlo commented:
Going through the selector / test flow manually (or running e2e in slow/headless mode) I do not see any issue, it seems to work as expected.

We do get an alert pop-up to confirm the change (confirming this is what makes the `editor-post-publish-button` selector available.  Maybe this is an issue where we need to add a `page.waitForSelector` before trying to click the button.... but it is odd that this is only surfacing now.  Im going to try that and see if it fixes Travis.  🤔

Addison-Stavlo committed:
added waitForSelector to failing post-visibility test

epiqueras commented:
`SCRIPT_DEBUG` is a PHP constant.

https://github.com/WordPress/gutenberg/tree/master/packages/env#wp-envjson

Addison-Stavlo commented:
>SCRIPT_DEBUG is a PHP constant.

Oh 😄  Thank you!  Now everything is passing locally for me both with and without the waitForSelector I added.  I'm hoping that makes Travis happy. 🤞 🤷‍♀️

Addison-Stavlo committed:
changed waitFor to wait for button to be clickable

Addison-Stavlo commented:
Nice, Travis passed on 66e669d !

I am updating that selector again though, to be sure we are waiting for the enabled button to be present.  Since the selector first appears with the button in the disabled state, this ensures the changes are registered and the button is clickable before attempting to click it.

MichaelArestad commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -6,32 +6,97 @@ import { some, groupBy } from 'lodash';
+						onClick={ selectParentBlock }
+						className="entities-saved-states__find-entity"
+					>
+						{ __( 'Select Entity' ) }
```
This is super interesting. It seems like a fun feature to experiment with. Perhaps the copy should just be "Select" for now.

Addison-Stavlo replied:
Thanks!  It was inspired by your mock-up 😄 - I just have it as a separate button temporarily since we are still using the checkbox controls for now.

> Perhaps the copy should just be "Select" for now.

I am agreeable to that, I could see a general using thinking "what the heck is an entity?".  Pushing that change now.

Addison-Stavlo committed:
changed 'Select entity' button to 'Select'

epiqueras commented on packages/base-styles/_z-index.scss:
```diff
@@ -69,7 +69,13 @@ $z-layers: (
+	".edit-post-layout .edit-post-post-publish-panel {greater than small}": 99998,
+
+	".entities-saved-states__panel": 100001,
+	// For larger views, the wp-admin navbar dropdown should be at top of
```
```suggestion
	// For larger views, the wp-admin navbar dropdown should be on top of
```

epiqueras commented on packages/e2e-tests/specs/experiments/multi-entity-saving.test.js:
```diff
@@ -75,7 +85,6 @@ describe( 'Multi-entity save flow', () => {
 		describe( 'Pre-Publish state', () => {
 			it( 'Should not trigger multi-entity save button with only post edited', async () => {
 				await createNewPost();
-				await disablePrePublishChecks();
```
Why was this removed?

Addison-Stavlo replied:
I don't I had a great reason when I added it first place (probably to keep things as simple as possible). 

 But it didn't seem to make sense anymore if we are testing the flow with the publish panel and expecting it to open after the save panel use through that flow, etc.

epiqueras commented on packages/e2e-tests/specs/experiments/multi-entity-saving.test.js:
```diff
@@ -101,14 +117,39 @@ describe( 'Multi-entity save flow', () => {
 
-			it( 'Saving should result in items being saved', async () => {
+			it( 'Should not show other panels (or their a11y buttons) while save panel opened', async () => {
+				await assertExistance( publishA11ySelector, false );
+				await assertExistance( saveA11ySelector, false );
+				await assertExistance( publishPanelSelector, false );
+			} );
+
+			it( 'Publish panel should open after saving, no other panels (or their a11y buttons) should be present', async () => {
+				// Save entities and wait for publish panel.
 				await page.click( entitiesSaveSelector );
+				await page.waitForSelector( publishPanelSelector );
+
+				await assertExistance( publishA11ySelector, false );
+				await assertExistance( saveA11ySelector, false );
+				await assertExistance( savePanelSelector, false );
```
Shouldn't we also check that the correct panel is open here?

Addison-Stavlo replied:
We are with the `waitForSelector`.  If the correct panel was not open, that would timeout and the test would fail.

epiqueras commented on packages/e2e-tests/specs/experiments/multi-entity-saving.test.js:
```diff
@@ -122,13 +163,17 @@ describe( 'Multi-entity save flow', () => {
 
 		describe( 'Published state', () => {
 			it( 'Update button disabled after publish', async () => {
-				await publishPostWithPrePublishChecksDisabled();
+				await publishPost();
```
Why was this changed?

Addison-Stavlo replied:
To properly publish the post now that I removed the unnecessary `disablePrePublishChecks`

epiqueras commented on packages/e2e-tests/specs/experiments/multi-entity-saving.test.js:
```diff
@@ -177,11 +235,20 @@ describe( 'Multi-entity save flow', () => {
 		} );
 
-		it( 'Clicking button should open modal with boxes checked', async () => {
+		it( 'save a11y button should be present', async () => {
+			await assertExistance( saveA11ySelector, true );
+		} );
+
+		it( 'Clicking button should open panel with boxes checked', async () => {
 			await page.click( activeSaveSiteSelector );
+			await page.waitForSelector( savePanelSelector );
 			await assertAllBoxesChecked();
 		} );
 
+		it( 'save a11y button should not be present with save panel open', async () => {
+			await assertExistance( saveA11ySelector, false );
+		} );
```
I think we should test all 4 selectors here. Something like `assertCorrectButtonOrPanelShowing` could help.

Addison-Stavlo replied:
This section is in the context of the site editor, where the publish panel / a11y button are currently non-existent.  I don't think it would make sense at this point in time, and if we were to implement the publish panel here similarly at some point in the future it will have different selectors.  I think it would make sense to update that when the time comes?

epiqueras replied:
But it's the same component, right? I meant to guard against it showing it when it shouldn't.

Addison-Stavlo replied:
> But it's the same component, right? 

Not really.  For edit-post we have the <ActionsPanel /> component that returns the 4.

In edit-site we aren't using that.  We just have a ternary statement controlling whether we show the 1 panel or the 1 button.  The publish panel / button aren't imported into anything in use.

Addison-Stavlo committed:
Update packages/base-styles/_z-index.scss

Co-Authored-By: Enrique Piqueras <epiqueras@users.noreply.github.com>

epiqueras commented on packages/edit-post/src/components/layout/actions-panel.js:
```diff
@@ -0,0 +1,93 @@
+				isOpen={ isEntitiesSavedStatesOpen }
+				closePanel={ closeEntitiesSavedStates }
+			/>
+			{ ! isEntitiesSavedStatesOpen && <>{ unmountableContent }</> }
```
```suggestion
			{ ! isEntitiesSavedStatesOpen && unmountableContent }
```

epiqueras commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -136,6 +129,26 @@ function Layout() {
 	}, [ isInserterOpen, isHugeViewport ] );
 
+	// Local state for save panel.
+	const [ isEntitiesSavedStatesOpen, setEntitiesSavedStates ] = useState(
+		false
+	);
+	const [
+		closeEntitiesCallback,
+		setCloseEntitiesCallback,
+	] = useState( () => () => {} );
```
Why not just hold the callback? It's truthy.

Addison-Stavlo replied:
It's an odd quirk of `useState`.  If we gave it `() => {}` it would accept `{}` as the initial value and on use we would get `closeEntitiesCallback is not a function`.    [Here are a couple explanations](https://stackoverflow.com/questions/55621212/is-it-possible-to-react-usestate-in-react) from the overflow.

epiqueras replied:
I meant, why don't we have a single state value?

```js
const [ entitiesSavedStatesCallback, setEntitiesSavedStatesCallback ] = useState();
```

`entitiesSavedStatesCallback` would be falsy when the panel should be closed, and hold the truthy function when it is open.

Addison-Stavlo replied:
The callback only gets set when used from the publish/update button.  When we go through the a11y button flow to open the save panel we don't need / don't set any callback.  So instead of checking to see if the callback is truthy before firing it, we initialize it as something that can handle being fired in the first place.

Do you think it would make more sense to initialize this as falsy and change the close function similarly:
```
	const closeEntitiesSavedStates = useCallback(
		( callbackArg ) => {
			closeEntitiesCallback && closeEntitiesCallback( callbackArg );
			setIsEntitiesSavedStatesOpen( false );
		},
		[ closeEntitiesCallback ]
	);
```
instead?

epiqueras replied:
`isEntitiesSavedStatesOpen` doesn't need to exist.

To open the panel without a callback, just set it to a no-op or `true` and then do a `typeof` check before calling it in `closeEntitiesSavedStates`.

Addison-Stavlo replied:
That makes sense.  I found an interesting edge case I had to fix while implementing this (although not related to doing it this way).  The fix is in 76e1f70 .

1.  I never reset the callback when closing the panel, this would mean that we could save through the 'publish' button flow, make edits, then use the save a11y flow that would try to use the same callback still and sometimes error.

2.  It seems that the `useCallback` for `dismissPanel` was causing the callback from the publish button to be skipped.  There would be no error, but the set callback would not fire and it would move on to just closing the panel instead.  This means the reset noted in 1 above would not happen when dismissing the panel from the close (X) button.

epiqueras commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -136,6 +129,26 @@ function Layout() {
 	}, [ isInserterOpen, isHugeViewport ] );
 
+	// Local state for save panel.
+	const [ isEntitiesSavedStatesOpen, setEntitiesSavedStates ] = useState(
```
```suggestion
	const [ isEntitiesSavedStatesOpen, setIsEntitiesSavedStatesOpen ] = useState(
```

epiqueras commented on packages/edit-site/src/components/editor/index.js:
```diff
@@ -68,6 +73,18 @@ function Editor( { settings: _settings } ) {
 
 	const inlineStyles = useResizeCanvas( deviceType );
 
+	const [ isEntitiesSavedStatesOpen, setEntitiesSavedStates ] = useState(
```
Same here.

epiqueras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -49,18 +114,22 @@ function EntityTypeList( { list, unselectedEntities, setUnselectedEntities } ) {
-		( select ) => select( 'core' ).__experimentalGetDirtyEntityRecords(),
-		[]
-	);
+export default function EntitiesSavedStates( { isOpen, closePanel } ) {
```
```suggestion
export default function EntitiesSavedStates( { isOpen, close } ) {
```

Would be cleaner.

Addison-Stavlo replied:
We have `close` in the upper scope as an icon import, we could import it under a different name but I see `closePanel` as being more semantic anyways.  As this gets updated in the future with more functionality, it may be required that other things are opened/closed and in that case `closePanel` will be more descriptive? 🤔

epiqueras replied:
Then `isOpen` should be `isPanelOpen`. But, I doubt we'll need that. I think it's reasonable to assume that the panel will be the main thing that is "open" and the non-suffixed props refer to that.

Addison-Stavlo replied:
Good points.  Updated!

epiqueras commented on packages/editor/src/components/post-publish-button/index.js:
```diff
@@ -49,6 +48,10 @@ export class PostPublishButton extends Component {
 					entitiesSavedStatesCallback: () => callback( ...args ),
 				} );
+				this.props.setCloseEntitiesCallback(
+					() => this.closeEntitiesSavedStates
```
```suggestion
					this.closeEntitiesSavedStates
```

It's already bound.

Addison-Stavlo replied:
Similar to the above, when setting a function in `useState`, if we were to only pass the function it would try to set that functions return as the value on state, erroring with `closeEntitiesCallback` is not a function.

epiqueras replied:
Right, because this is the actual setter, not a proxy. A comment to clarify would be useful.

epiqueras reviewed (approved):
Nice work 😄

Addison-Stavlo committed:
applying some suggestions

Addison-Stavlo committed:
Update packages/edit-post/src/components/layout/actions-panel.js

Co-Authored-By: Enrique Piqueras <epiqueras@users.noreply.github.com>

Addison-Stavlo committed:
added comment to setting callback

Addison-Stavlo committed:
updated callback goo

epiqueras commented on packages/editor/src/components/post-publish-button/index.js:
```diff
@@ -80,6 +80,8 @@ export class PostPublishButton extends Component {
 			}
 		} );
+		// Clear the callback set for entities saved states.
+		this.props.setCloseEntitiesCallback();
```
This is the responsibility of `closeEntitiesSavedStates`.

Addison-Stavlo replied:
Agreed.  This is the last line of `closeEntitiesSavedStates`.

epiqueras replied:
I meant that the callback should be cleared here https://github.com/WordPress/gutenberg/blob/9ea1a04178dc5d736b18f363e4f78bc0535143e6/packages/edit-post/src/components/layout/index.js#L142-L150.

Not by the caller. It's something that always should happen after calling the callback. Not something that the caller should decide on.

Addison-Stavlo replied:
Ah ok, that makes sense!  

So instead of relying on `closeEntitiesCallback` to have been defined in a way that resets the callback after use, we just clear the callback after calling it.  Thanks again =) 

This is updated.

epiqueras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -180,7 +180,9 @@ export default function EntitiesSavedStates( { isOpen, closePanel } ) {
-	const dismissPanel = useCallback( () => closePanel(), [] );
+	// Do NOT useCallback, it will skip the 'this' bound callback used in the
+	// closePanel function.
+	const dismissPanel = () => closePanel();
```
I don't think that's the issue here.

You are closing over `closePanel` which changes when the callback changes, so you need to add it as a dep.

Addison-Stavlo replied:
Ohhh yes!  100%.  So it was firing `closePanel` as it was at the state on mount, before the callback was set.  That makes sense.  Thank you!

Addison-Stavlo replied:
I am much happier now that the issue was the dep.  I was a bit confused late last night 😆 .

This should be fixed now as well.

chrisvanpatten commented:
This is so sweet. It would be cool to see reusable blocks move to a similar saving approach. Nice work here, can't wait to use it! 🙌

Addison-Stavlo committed:
fixed useCallback for dismissing panel

Addison-Stavlo committed:
updates prop name from closePanel to close

noahtallen commented:
> It would be cool to see reusable blocks move to a similar saving approach

Conveniently, #21368 sets us up for something like that 👀

noahtallen reviewed (approved):
Let's do it! :shipit: 

I noticed a thing where an edit to a template part _after_ an initial save did not make the editor think that the TP had changes. so it doesn't persist the change to the DB when you click save. I'm going to assume that this will be fixed by #21368, since we already know that dirty states aren't being computed correctly. 😁 But I thought I'd point it out because I didn't think this was a bug on master

<img width="1882" alt="Screen Shot 2020-04-30 at 9 27 04 AM" src="https://user-images.githubusercontent.com/6265975/80734944-d2053080-8ac4-11ea-8ea1-0fe3badbca81.png">

Addison-Stavlo committed:
updated how we clear the callback

Addison-Stavlo commented:
> I noticed a thing where an edit to a template part after an initial save did not make the editor think that the TP had changes. 

@noahtallen, I'm not able to reproduce this.  I tried both inserting an existing TP as well as creating a new one.  The TP always seems to come up as 'dirty' after changes are made to it on my end.

noahtallen commented:
no worries :) I don't think it should affect this PR then. If it comes up again, it would probably be related to the code that I'm working on anyways rather than the UI aspect of displaying it

Addison-Stavlo commented:
I think all the requested changes have been addressed.  Are we in a good state to move forward with this now?

epiqueras commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -136,6 +129,27 @@ function Layout() {
 		}
 	}, [ isInserterOpen, isHugeViewport ] );
 
+	// Local state for save panel.
+	const [
+		isEntitiesSavedStatesOpen,
+		setIsEntitiesSavedStatesOpen,
+	] = useState( false );
+	const [ closeEntitiesCallback, setCloseEntitiesCallback ] = useState();
+	const openEntitiesSavedStates = useCallback(
+		() => setIsEntitiesSavedStatesOpen( true ),
+		[]
+	);
+	const closeEntitiesSavedStates = useCallback(
+		( callbackArg ) => {
+			if ( closeEntitiesCallback ) {
+				closeEntitiesCallback( callbackArg );
+				setCloseEntitiesCallback();
+			}
+			setIsEntitiesSavedStatesOpen( false );
+		},
+		[ closeEntitiesCallback ]
+	);
+
```
```js
const [
	entitiesSavedStatesCallback,
	setEntitiesSavedStatesCallback,
] = useState( false );
const closeEntitiesSavedStates = useCallback(
	( arg ) => {
		if ( typeof entitiesSavedStatesCallback === 'function' ) {
			entitiesSavedStatesCallback( arg );
		}
		setEntitiesSavedStatesCallback( false );
	},
	[ entitiesSavedStatesCallback ]
);
```

Non-blocking, but it would be simpler.

```js
if ( entitiesSavedStatesCallback ) {
	// It's open.
}

// Open with callback.
setEntitiesSavedStatesCallback( () => () => {} );

// Just open.
setEntitiesSavedStatesCallback( true );
```

Addison-Stavlo replied:
That is really neat!  Thank you for the suggestion, I will apply this as it does seem to clean up a couple things.

vindl reviewed (approved):
This worked as expected in my testing. Nice work @Addison-Stavlo! 🎉 

I noticed a minor issue with `Select` functionality. Clicking it for the second time will de-select the template part, and consecutive clicks won't toggle it back. However I don't think it's a blocker and it can be addressed in follow-up PR if needed.

Addison-Stavlo committed:
simplified logic on callback setting and open/close

Addison-Stavlo committed:
minor comment nit

youknowriad commented on packages/editor/src/components/entities-saved-states/style.scss:
```diff
@@ -1,11 +1,70 @@
-		font-size: 18px;
-		margin: 20px 0 10px;
+.entities-saved-states__panel {
+	@include reset;
```
Do you recall why we need a "reset" here. Seems weird since this is meant for global resets.

youknowriad replied:
Ideally, components don't depend on global resets though.

Addison-Stavlo replied:
It may not be necessary.  I think I added it b/c post-publish-panel was using it at the time as well and I was trying to ensure the styles were the same between the two.  It looks like it's not there anymore either.
